designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/provider/ProcessingTypeDataProvider.scala

```diff
@@ -1,56 +1,75 @@
 import pl.touk.nussknacker.security.Permission
 import pl.touk.nussknacker.ui.UnauthorizedError
 import pl.touk.nussknacker.ui.process.processingtype.ValueWithRestriction
 import pl.touk.nussknacker.ui.security.api.LoggedUser
 
-import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.CopyOnWriteArrayList
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Try}
 
 /**
   * ProcessingType is a context of application. One ProcessingType can't see data from another ProcessingType.
   * Services run inside one ProcessingType scope behave differently from services run in another scope.
   *
   * This class is meant to provide access to some scope of data inside context of application to the user.
   * We don't want to pass all ProcessingType's data to every service because it would complicate testing of services
-  * and would broke isolation between areas of application. Due to that, this class is a `Functor`
+  * and would brake isolation between areas of application. Due to that, this class is a `Functor`
   * (to be precise `BiFunctor` but more on that below) which allows to transform the scope of `Data`.
   *
   * Sometimes it is necessary to have access also to combination of data across all ProcessingTypes. Due to that
   * this class is a `BiFunctor` which second value named as `CombinedData`
   *
   * This class caches `Data` and `CombinedData` wrapped in `ProcessingTypeDataState` to avoid computations of
-  * transformations during each lookup to `Data`/`CombinedData`. It behave similar to `Observable` where given
-  * transformed `ProcessingTypeDataProvider` check its parent if `ProcessingTypeDataState.stateIdentity` changed.
+  * transformations during each lookup to `Data`/`CombinedData`.
   *
   * ProcessingType is associated with Category e.g. Fraud Detection, Marketing. Given user has access to certain
-  * categories see `LoggedUser.can`. Due to that, during each access to `Data`, user is authorized if he/she
-  * has access to category.
+  * categories see `LoggedUser.can`. Due to that, during each access to `Data`, user is authorized if they
+  * have access to category.
   */
-trait ProcessingTypeDataProvider[+Data, +CombinedData] {
+abstract class ProcessingTypeDataProvider[Data, CombinedData](initialState: ProcessingTypeDataState[Data, CombinedData])
+    extends LazyLogging {
+
+  // We use unsafeRunSync because IO-based solution would complicate a lot our application initialization logic where
+  // we pass to every service, various views for our ProcessingTypeData
+  protected val stateRef: Ref[IO, ProcessingTypeDataState[Data, CombinedData]] =
+    Ref.of[IO, ProcessingTypeDataState[Data, CombinedData]](initialState).unsafeRunSync()(IORuntime.global)
+
+  protected val stateMutex: Mutex[IO] = Mutex[IO].unsafeRunSync()(IORuntime.global)
+
+  private val observers = new CopyOnWriteArrayList[Observer[ProcessingTypeDataState[Data, CombinedData]]]()
```

Comment:
Thanks to this `observers` based approach, it was possible to remove `ProcessingTypeDataState.stateIdentity`. Now state in `ProcessingTypeData` views, created by `.mapValues`/`.mapCombined` are reloaded eagerly instead of lazy.

Reply:
Looking from a perspective of `AkkaHttpBasedRouteProvider` class this is ok. But the perspective of the `ProcessingTypeDataProvider` class is different. This class accumulates some state and should clean it up too. Without taking into consideration its current usage (I mean: the class determines what it does, how it does, and how its state is managed, not its clients).

I'd propose to remove the observers in its close method. The behaviour obviously won't change but the solution is cleaner.